### PR TITLE
android: add automatic display on and use in benchmarks

### PR DESCRIPTION
**Pull request by joelagnel**
- Add Screen on/off and unlock support
- Use it in uibench to unlock and turn on screen

Signed-off-by: Joel Fernandes <joelaf@google.com>

### CODE REVIEW

**Comment by bjackman** [comment]
Would prefer one commit for libs/utils/android and one commit for UiBench. Other than that LGTM, but will wait for review from @ionela-voinescu or @derkling as they are more familiar with this code.

**Comment by joelagnel** [comment]
Guys, another comments here?

File: libs/utils/android/screen.py [hunk 1]
```diff
@@ -136,4 +137,25 @@ def get_screen_density(target):
         """
         return target.execute('getprop ro.sf.lcd_density')
 
+    @staticmethod
+    def is_screen_on(target):
+	cmdout = target.execute('dumpsys input_method')
+	return ('mInteractive=true' in cmdout)
+
+    @staticmethod
+    def set_screen_on(target):
```
Comment (derkling) [inline libs/utils/android/screen.py]: Moreover, I've found these two guys:
https://developer.android.com/reference/android/view/KeyEvent.html#KEYCODE_WAKEUP
https://developer.android.com/reference/android/view/KeyEvent.html#KEYCODE_SLEEP

Thus perhaps we can avoid the "expensive" check with dumpsys and go straight with these key events depending on the value of the `on` parameter.

File: libs/utils/android/screen.py [hunk 2]
```diff
@@ -136,4 +137,25 @@ def get_screen_density(target):
...
+        log = logging.getLogger('Screen')
+        if Screen.is_screen_on(target) == False:
+            log.info('Screen is off, turning on')
+            System.power(target)
+
+    @staticmethod
+    def unlock(target):
+       Screen.set_screen_on(target)
+       target.execute('input keyevent 82')
```
Comment (derkling) [inline libs/utils/android/screen.py]: This is `System.menu()`

File: libs/utils/android/screen.py [hunk 3]
```diff
@@ -136,4 +137,25 @@ def get_screen_density(target):
...
+       # Incase screen was already unlocked, code 82
+       # may cause it zoom out, in this case just issue
+       # back key to restore it
+       System.back(target)
```
Comment (derkling) [inline libs/utils/android/screen.py]: Perhaps this is not working in case the scree is already on and an app runnning, we are not granted to be in the home screen.

I think this sequence should be more robust:
```python
System.set_screen(on=True)
System.menu()
System.home()
```

This should grant to unlock the device and move back to the home screen... at least it did for all the starting conditions I've tested.

**Comment by joelagnel** [comment]
Patrick, thanks for the great review, I made changes, please review and merge, thanks

File: libs/utils/android/screen.py [hunk 4]
```diff
@@ -136,4 +137,20 @@ def get_screen_density(target):
         """
         return target.execute('getprop ro.sf.lcd_density')
 
+    @staticmethod
+    def set_screen(target, on=True):
+        log = logging.getLogger('Screen')
+        if on == True:
+            target.execute('input keyevent KEYCODE_WAKE')
```
Comment (derkling) [inline libs/utils/android/screen.py]: I would really like to have this keyevent (and the following one) exposed as System APIs.
Can you add these two calls to System in a patch before this one?

```python
System.wake()
System.sleep()
```

File: libs/utils/android/screen.py [hunk 5]
```diff
@@ -136,4 +137,20 @@ def get_screen_density(target):
...
+            log.info('Setting screen ON')
+        else:
```
Comment (derkling) [inline libs/utils/android/screen.py]: I would prefer:

```python
if not on:
    System.sleep()
    log.info('Setting screen OFF')
    return

System.wake()
log.info('Setting screen ON')
```

Which better match the API signature where also the non default case returns immediately.

File: libs/utils/android/workloads/uibench.py [hunk 6]
```diff
@@ -85,7 +85,7 @@ def run(self, out_dir, test_name, duration_s, collect=''):
         self.out_dir = out_dir
         self.collect = collect
 
-        # Press Back button to be sure we run the video from the start
+        Screen.unlock(self._target)
         System.menu(self._target)
```
Comment (derkling) [inline libs/utils/android/workloads/uibench.py]: A think you can make even more nicely red coloured this patch by removing this and the next call... aren't they already done by the new `Screen.unlock()` call?

Comment (derkling) [inline libs/utils/android/workloads/uibench.py]: Can you also please check if there are similar patterns in the other `libs/utils/android/workloads/*.py` and do the same change in all of them in this patch (which will became a more generic "android: use a more generic display unlocking function"

File: libs/utils/android/screen.py (hunk 4 above)
Comment (joelagnel) [inline libs/utils/android/screen.py]: I thought you preferred to send the event directly from Screen, I can certainly add a different path with that for System,... actually that's what I was going to do too.

File: libs/utils/android/screen.py (hunk 5 above)
Comment (joelagnel) [inline libs/utils/android/screen.py]: Ok.

File: libs/utils/android/workloads/uibench.py (hunk 6 above)
Comment (joelagnel) [inline libs/utils/android/workloads/uibench.py]: Re: First comment, there's some weirdness where if I don't do it in unlock() then the screen doesn't unlock. I didn't get too much time last time around to fix that, but I will investigate more today.

Re: Second comment, I normally don't push changes without testing them so I didn't change what I wasn't using ;-) although, now that you mentioned it, I will try to include it in this patch with some amount of testing

Comment (joelagnel) [inline libs/utils/android/workloads/uibench.py]: Found the reason, and sorry for the sloppy job. Fixing it, cheers

Comment (joelagnel) [inline libs/utils/android/workloads/uibench.py]: About second comment, I will submit the other workload changes in other patch series please, so that this can be merged first. thanks

**Comment by joelagnel** [comment]
Hope this set is better, thanks so much for your thorough review :-) :+1:

File: libs/utils/android/workloads/youtube.py [hunk 7]
```diff
@@ -66,9 +66,10 @@ def run(self, out_dir, video_url, video_duration_s, collect=''):
         self.collect = collect
 
         # Unlock device screen (assume no password required)
-        System.menu(self._target)
-        # Press Back button to be sure we run the video from the start
-        System.back(self._target)
+        Screen.unlock(self._target)
+
+        # Stop youtube if already running
+        System.force_stop(self._target, self.package)
```
Comment (derkling) [inline libs/utils/android/workloads/youtube.py]: Should not this better go before the unlock?

However, I would prefer to keep this chunk into a separate patch, which follow the one updating the unlock procedure. At the end that's required just by Youtube and to grant that we start from the beginning.

Comment (joelagnel) [inline libs/utils/android/workloads/youtube.py]: I don't think it matters whether it is before or after, so I prefer to leave this as it is in the patch.

**Comment by joelagnel** [comment]
Patrick? I pushed an update.
